contents/staticAnalysis/FindBugs.md

```diff
@@ -0,0 +1,59 @@
+# FindBugs
+
+Author: [Xiao Pu](https://nus-oss.github.io/cs3281-website/students/AY1617S2/xiaoPu/xiaoPu-Resume.html)
+
+## Overview
+
+FindBugs is a static analysis to find "bugs" in **Java** programme. It looks for "bug patterns" in code and signals possible violations.
+
+## Features
+
+The "bug patterns" can be divided into nine groups: Bad practice, Correctness, Experimental, Internationalization, Malicious code vulnerability, Multithreaded correctness, Performance, Security and Dodgy code. [A comprehensive list](http://findbugs.sourceforge.net/bugDescriptions.html) of bugs is provided to explain the meaning of each bug.
+
+FindBugs analyses code based on compiled Java `.class` file and checks multiple files at the same time. These feature overcomes some limitations in [CheckStyle](CheckStyle.md) and [PMD](PMD.md). Therefore, FindBugs can spot potential errors that CheckStyle and PMD can not find. For example, one of the bug patterns in FindBugs is `RCN: Redundant nullcheck of value known to be non-null`. Only after a comprehensive analysis of all the assignments to a variable in the code base, we can check whether the `nullcheck` of value is redundant or not.
```

Comment:
... analyses **bytecode**

_Probably need a bit of rephrasing because the current version sounds like it analysis source code rather than byte code._